### ISSUE TITLE
GPU Support

### ISSUE THREAD

**Issue by Ulfgard** [Shark-ML/Shark#218]
With pull #217 in place, and the state of development of Remora, we can start pushing the first Algorithms to GPU. We should start with something small like(Logistic) Regression via ErrorFunction with CrossEntropy/SquaredLoss using gradient descent. This requires:

- [ ] Move a Dataset to GPU. This is straight forward for vector data as we can simply map the CPU matrices to GPU matrices using remora::move_to_gpu. It becomes harder with class labels. Maybe we should convert them to one-hot encoding for the time being
- [ ] Ensure the LinearModel is GPU capable. Here, serialization might make our life hard. The rest should work.
- [ ] Check that SquaredLoss is GPU capable. should be straight forward
- [ ] Port CrossEntropy, This is a bigger challenge given the current implementation
- [ ]  Error Function should be straight forward. Threading might have to be disabled on GPU. but mini batches should work out of the box.
- [ ] SteepestDescent should work out of the box. Again, serialization.

if all this works: port LBFGS and try to run the full LogisticRegression trainer.

**Comment by Ulfgard** [comment]
Status: I got the Autoencoder example to run on GPU. However, a canonical code to move datasets is still missing. Working on that next. @kremerj

**Comment by Ulfgard** [comment]
Things needed in Remora for an efficient and stable implementation of cross-entropy with vectorial/probabilitylabels:
[ ] maximum of each row of a matrix
[ ] row-wise scalar-product between two matrices

**Comment by Ulfgard** [comment]
LBFGS is maybe a bit too optimistic, it probably requires rewriting the algorithm in terms of the block-update. However, we can now get a lot of stuff working on GPU. 

We can now train neural networks for regression and classification. There is still a bit work on the convolution, however this probably depends on the new clBLAST conv kernel. So this can be done later.

Issues:
- ZeroOneLoss 
- Classifier<Model>
- I think we should make the transfer to one-hot explicit as we might want to change this later.
- ConvNets need to be ported
- very big overhead for creating kernels through boost.compute. Kernels should be cached as well.
